client/core/core.go

```diff
@@ -4260,64 +4266,71 @@ out:
 }
 
 // handlePreimageRequest handles a DEX-originating request for an order
-// preimage.
+// preimage. If the order id in the request is not known, it may launch a
+// goroutine to wait for a market/limit/cancel request to finish processing.
 func handlePreimageRequest(c *Core, dc *dexConnection, msg *msgjson.Message) error {
 	req := new(msgjson.PreimageRequest)
 	err := msg.Unmarshal(req)
 	if err != nil {
 		return fmt.Errorf("preimage request parsing error: %v", err)
 	}
 
+	if len(req.OrderID) != order.OrderIDSize {
+		return fmt.Errorf("invalid order ID in preimage request")
+	}
+
 	var oid order.OrderID
 	copy(oid[:], req.OrderID)
 
-	// Sync with order placement.
+	// Sync with order placement, the response from which provides the order ID.
 	c.piSyncMtx.Lock()
-	syncChan, found := c.piSyncers[oid]
-	if found {
-		// If we found the channel, the tracker is already in the map, and we
-		// can go ahead.
-		delete(c.piSyncers, oid)
-	} else {
-		// If we didn't find the channel, Trade is still running. Add the chan
-		// and wait for Trade to close it.
-		syncChan = make(chan struct{})
-		c.piSyncers[oid] = syncChan
-	}
-	c.piSyncMtx.Unlock()
-
-	deletePiSyncer := func() {
-		c.piSyncMtx.Lock()
+	if _, found := c.piSyncers[oid]; found {
+		// If we found a map entry, the tracker is already in the trades map,
+		// and we can go ahead.
 		delete(c.piSyncers, oid)
 		c.piSyncMtx.Unlock()
+		return processPreimageRequest(c, dc, msg.ID, oid)
 	}
 
-	if !found {
+	// If we didn't find an entry, Trade or Cancel is still running. Insert a
+	// chan and wait for it to close the channel.
+	syncChan := make(chan struct{})
+	c.piSyncers[oid] = syncChan
+	c.piSyncMtx.Unlock()
+
+	// The order submission could be timing out waiting for a response, or this
+	// could be a bogus preimage request, so we do not want to block the caller,
+	// (*Core).listen if this hangs. Preimage requests are ok to handle
+	// asynchronously since there can be no matches until we respond to this.
+	c.log.Warnf("Got preimage request for %v before we got an order submission response!", oid)
+	go func() {
 		select {
 		case <-syncChan:
-		case <-time.After(time.Minute):
-			deletePiSyncer()
-			return fmt.Errorf("timed out syncing preimage request for %s, order %s", dc.acct.host, oid)
```

Comment:
This was the trouble.  `prepareTrackedTrade` and `tryCancel` could return before getting to `syncOrderPlaced`, but with server having actually received the order and thus sent this preimage request that client cannot handle.